modules/optimization/src/reporters/ParsedReporterBase.C

```diff
@@ -0,0 +1,88 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ParsedReporterBase.h"
+
+// registerMooseObject("OptimizationApp", ParsedReporterBase);
+
+InputParameters
+ParsedReporterBase::validParams()
+{
+  InputParameters params = GeneralReporter::validParams();
+  params += FunctionParserUtils<false>::validParams();
+  params.addRequiredCustomTypeParam<std::string>(
+      "expression", "FunctionExpression", "function expression");
+  params.addParam<std::string>("name", "result", "Name of output reporter.");
+  params.addRequiredParam<std::vector<std::string>>("reporter_symbols",
+                                                    "Expression symbol for each reporter");
+  params.addParam<std::vector<std::string>>(
+      "constant_names",
+      {},
+      "Vector of constants used in the parsed function (use this for kB etc.)");
+  params.addParam<std::vector<std::string>>(
+      "constant_expressions",
+      {},
+      "Vector of values for the constants in constant_names (can be an FParser expression)");
+  params.addParam<bool>(
+      "use_t", false, "Make time (t) variables available in the function expression.");
+  return params;
+}
+
+ParsedReporterBase::ParsedReporterBase(const InputParameters & parameters)
+  : GeneralReporter(parameters),
+    FunctionParserUtils(parameters),
+    _use_t(getParam<bool>("use_t")),
+    _reporter_symbols(getParam<std::vector<std::string>>("reporter_symbols"))
+{
+  // build reporters argument
+  std::string symbol_str;
+  for (const auto i : index_range(_reporter_symbols))
+    symbol_str += (i == 0 ? "" : ",") + _reporter_symbols[i];
+
+  // add time if required
+  if (_use_t)
+    symbol_str += (symbol_str.empty() ? "" : ",") + std::string("t");
+
+  // base function object
+  _func_F = std::make_shared<SymFunction>();
+
+  // set FParser internal feature flags
+  setParserFeatureFlags(_func_F);
+
+  // add the constant expressions
+  addFParserConstants(_func_F,
+                      getParam<std::vector<std::string>>("constant_names"),
+                      getParam<std::vector<std::string>>("constant_expressions"));
+
+  // parse function
+  std::string function = getParam<std::string>("expression");
+  if (_func_F->Parse(function, symbol_str) >= 0)
+    mooseError("Invalid parsed function\n", function, "\n", _func_F->ErrorMsg());
+
+  // optimize
+  if (!_disable_fpoptimizer)
+    _func_F->Optimize();
+
+  // just-in-time compile
+  if (_enable_jit)
+  {
+    // let rank 0 do the JIT compilation first
+    if (_communicator.rank() != 0)
+      _communicator.barrier();
+
+    _func_F->JITCompile();
+
+    // wait for ranks > 0 to catch up
+    if (_communicator.rank() == 0)
+      _communicator.barrier();
```

Comment:
@dschwen is this how we do this elsewhere?

Reply:
Most of this constructor is copied from ParsedAux.

Reply:
👍🏼